gobblin-core/src/main/java/org/apache/gobblin/writer/PartitionedDataWriter.java

```diff
@@ -99,13 +118,32 @@ public PartitionedDataWriter(DataWriterBuilder<S, D> builder, final State state)
     if(builder.schema != null) {
       this.state.setProp(WRITER_LATEST_SCHEMA, builder.getSchema());
     }
-    this.partitionWriters = CacheBuilder.newBuilder().build(new CacheLoader<GenericRecord, DataWriter<D>>() {
+    Long cacheExpiryInterval = this.state.getPropAsLong(PARTITIONED_WRITER_CACHE_TTL_SECONDS, DEFAULT_PARTITIONED_WRITER_CACHE_TTL_SECONDS);
+
+    this.partitionWriters = CacheBuilder.newBuilder()
+        .expireAfterAccess(cacheExpiryInterval, TimeUnit.SECONDS)
+        .removalListener(new RemovalListener<GenericRecord, DataWriter<D>>() {
+      @Override
+      public void onRemoval(RemovalNotification<GenericRecord, DataWriter<D>> notification) {
+        synchronized (PartitionedDataWriter.this) {
+          if (notification.getValue() != null) {
+            try {
+              DataWriter<D> writer = notification.getValue();
+              totalRecordsFromEvictedWriters += writer.recordsWritten();
+              totalBytesFromEvictedWriters += writer.bytesWritten();
+              writer.close();
+            } catch (IOException e) {
+              log.error("Exception {} encountered when closing data writer on cache eviction", e);
```

Comment:
Does failure need to be propagated to fail the task, like if a close failed?

Reply:
Hmm. I was not sure if we should make close exceptions fatal. I am propagating the exception per the suggestion. Just curious if there are strong reasons we want to make close exceptions fatal.

Reply:
The existing close that happens from the closer propagates errors. There are some writers like the `HiveWritableHdfsDataWriter` that finalizes files on close. If that fails then there may be an incomplete file and the task should fail instead of publish.